zaza/openstack/charm_tests/ceph/tests.py

```diff
@@ -749,138 +895,231 @@ def test_services(self):
+        """
+        logging.info('Checking Object Storage')
+        container_name = 'zaza-container'
+        obj_data = 'Test data from Zaza'
+
+        # Create client for IO with Primary endpoint.
+        primary_endpoint = self.get_rgw_endpoint(self.primary_rgw_unit)
+        secondary_endpoint = self.get_rgw_endpoint(self.secondary_rgw_unit)
+        self.assertNotEqual(primary_endpoint, None)
+        self.assertNotEqual(secondary_endpoint, None)
+
+        access_key, secret_key = self.get_client_keys() 
+        primary_client = boto3.resource("s3",
+                                        verify=False,
+                                        endpoint_url=primary_endpoint,
+                                        aws_access_key_id=access_key,
+                                        aws_secret_access_key=secret_key)
+        primary_client.Bucket(container_name).create()
+        primary_object_one = primary_client.Object(
+            container_name,
+            'prefile'
+        )
+        primary_object_one.put(Body=obj_data)
+        content = primary_object_one.get()['Body'].read().decode('UTF-8')
 
-        try:
-            source_client.head_object(_container, 'testfile')
-        except ClientException as e:
-            self.assertEqual(e.http_status, 404)
-        else:
-            self.fail('object not deleted on source radosgw')
+        # 1. Verify object storage API works as expected.
+        self.assertEqual(content, obj_data)
+        logging.info('Object Storage OK')
+
+        # Migrate to multisite if multisite-model is deployed.
+        if not self.multisite:
+            logging.info('Skipping Multisite Migration Test')
+            return
 
-    def test_multisite_failover(self):
+        logging.info('Checking Multisite Migration')
+        # If Master/Slave relation does not exist.
+        if zaza_model.get_relation_id(
+                self.primary_rgw_app, self.secondary_rgw_app,
+                remote_interface_name='slave'
+        ) is None:
+            logging.info('Configuring Multisite')
+            self.configure_rgw_apps_for_multisite()
+            zaza_model.add_relation(
+                self.primary_rgw_app,
+                self.primary_rgw_app + ":master",
+                self.secondary_rgw_app + ":slave"
+            )
+            zaza_model.block_until_unit_wl_status(
+                self.secondary_rgw_unit, "waiting"
+            )
+
+        zaza_model.block_until_unit_wl_status(
+            self.secondary_rgw_unit, "active"
+        )
+        logging.info('Waiting for Data and Metadata to Synchronize')
+        self.wait_for_sync(self.secondary_rgw_app, isPrimary=False)
+        self.wait_for_sync(self.primary_rgw_app, isPrimary=True)
+
+        # Create a new session and Add one more object on primary.
+        logging.info('Performing post migration IO tests.')
+        primary_client = boto3.resource("s3",
```

Comment:
Is there some reason we can't use the existing `primary_client`?

Reply:
Sometimes when there is a significant delay between subsequent requests it appears as though some internal timeout occurs and existing boto3 resource objects (Bucket instances and RGW Object Instances) can't be reused (have to be re-instantiated). It appears that doesnt happen for client instance, I have removed this for now. If it breaks in further testing I will place it back again.